presto-analyzer/src/main/java/com/facebook/presto/sql/analyzer/QueryPreparer.java

```diff
@@ -58,27 +55,29 @@ public QueryPreparer(SqlParser sqlParser)
         this.sqlParser = requireNonNull(sqlParser, "sqlParser is null");
     }
 
-    public PreparedQuery prepareQuery(Session session, String query, WarningCollector warningCollector)
+    public PreparedQuery prepareQuery(AnalyzerOptions analyzerOptions, String query, Map<String, String> preparedStatements, WarningCollector warningCollector)
             throws ParsingException, PrestoException, SemanticException
     {
-        Statement wrappedStatement = sqlParser.createStatement(query, createParsingOptions(session, warningCollector));
-        if (warningCollector.hasWarnings() && getWarningHandlingLevel(session) == AS_ERROR) {
+        Statement wrappedStatement = sqlParser.createStatement(query, analyzerOptions.getParsingOptions());
+        if (warningCollector.hasWarnings() && analyzerOptions.getWarningHandlingLevel() == AS_ERROR) {
             throw new PrestoException(WARNING_AS_ERROR, format("Warning handling level set to AS_ERROR. Warnings: %n %s",
                     warningCollector.getWarnings().stream()
                             .map(PrestoWarning::getMessage)
                             .collect(joining(System.lineSeparator()))));
         }
-        return prepareQuery(session, wrappedStatement, warningCollector);
+        return prepareQuery(analyzerOptions, wrappedStatement, preparedStatements);
     }
 
-    public PreparedQuery prepareQuery(Session session, Statement wrappedStatement, WarningCollector warningCollector)
+    public PreparedQuery prepareQuery(AnalyzerOptions analyzerOptions, Statement wrappedStatement, Map<String, String> preparedStatements)
             throws ParsingException, PrestoException, SemanticException
     {
         Statement statement = wrappedStatement;
         Optional<String> prepareSql = Optional.empty();
         if (statement instanceof Execute) {
-            prepareSql = Optional.of(session.getPreparedStatementFromExecute((Execute) statement));
-            statement = sqlParser.createStatement(prepareSql.get(), createParsingOptions(session, warningCollector));
+            String preparedStatementName = ((Execute) statement).getName().getValue();
+            prepareSql = Optional.ofNullable(preparedStatements.get(preparedStatementName));
```

Comment:
lambdas and optional feel superfluous here. A simple if (statement != null)
...
else {
   throw

would likely be more readable

Reply:
prepareSql is part of the PreparedQuery, which gets used in many places in the engine. So we need to populate it. I also think Optional is a good container to pass it around as it helps in avoiding null checking.